src/Scriban/Runtime/Accessors/TypedObjectAccessor.cs

```diff
@@ -95,20 +95,32 @@ public bool TrySetItem(TemplateContext context, SourceSpan span, object target,
 
         public bool HasIndexer => _indexer != null;
 
+        public static object ChangeType(object value, Type targetType)
+        {
+            if (value == null)
+                return null;
+
+            if (value.GetType() == targetType)
+                return value;
+
+            if (targetType.IsGenericType && targetType.GetGenericTypeDefinition() == typeof(Nullable<>))
+                targetType = Nullable.GetUnderlyingType(targetType);
+
+            return Convert.ChangeType(value, targetType);
+        }
+
         public bool TrySetValue(TemplateContext context, SourceSpan span, object target, string member, object value)
         {
-            MemberInfo memberAccessor;
-            if (_members.TryGetValue(member, out memberAccessor))
+            if (_members.TryGetValue(member, out MemberInfo memberAccessor))
             {
-                var fieldAccessor = memberAccessor as FieldInfo;
-                if (fieldAccessor != null)
+                if (memberAccessor is FieldInfo fieldAccessor)
                 {
-                    fieldAccessor.SetValue(target, value);
+                    fieldAccessor.SetValue(target, ChangeType(value, fieldAccessor.FieldType));
```

Comment:
Conversion between value to a new type should use `context.ToObject` instead. `ChangeType` should not be introduced

Reply:
Righto. Updated